src/lookup.ts

```diff
@@ -168,61 +179,64 @@ class Lookup {
       object[p] = array[lookup[p] - offset];
     }
     return object;
-  }
-
-  /**
-   *
-   * @param {Array} array
-   * @returns {*}
-   */
-  static lookupFromArray(array) {
-    const lookup = {};
-    let z = 0;
-    let i = array.length;
-    while (i-- > 0) {
-      lookup[array[i]] = z++;
-    }
-    return lookup;
-  }
+  },
 
-  static dataShape(data) {
+  dataShape(
+    data: ITrainingData | ITrainingData[] | Datum | Datum[] | Datum[][]
+  ): string[] {
     const shape = [];
-
-    if (data.input) {
+    let lastData: Datum | Datum[] | Datum[][];
+    if (data.hasOwnProperty('input')) {
       shape.push('datum');
-      data = data.input;
+      lastData = (data as ITrainingData).input;
     } else if (Array.isArray(data)) {
-      if (data[0].input) {
+      if ((data as ITrainingData[])[0].input) {
         shape.push('array', 'datum');
-        data = data[0].input;
-      } else {
+        lastData = (data as ITrainingData[])[0].input;
+      } else if (Array.isArray(data[0])) {
         shape.push('array');
-        data = data[0];
+        lastData = data[0];
+      } else {
+        lastData = data as Datum | Datum[] | Datum[][];
       }
+    } else {
+      lastData = data as Datum | Datum[] | Datum[][];
     }
 
     let p;
-    while (data) {
-      for (p in data) {
-        break;
-      }
-      if (!data.hasOwnProperty(p)) break;
-      if (Array.isArray(data) || data.buffer instanceof ArrayBuffer) {
+    while (lastData) {
+      p = Object.keys(lastData)[0];
+      if (Array.isArray(lastData) || lastData.buffer instanceof ArrayBuffer) {
         shape.push('array');
-        data = data[p];
-      } else if (typeof data === 'object') {
+        const possibleNumber:
+          | number
+          | IArrayDatum = (lastData as IArrayDatum[])[parseInt(p)];
+        if (typeof possibleNumber === 'number') {
+          shape.push('number');
+          break;
+        } else {
+          lastData = possibleNumber;
+        }
+      } else if (typeof lastData === 'object') {
         shape.push('object');
-        data = data[p];
+        const possibleNumber:
+          | number
+          | IObjectDatum = (lastData as IObjectDatum)[p];
+        if (typeof possibleNumber === 'number') {
+          shape.push('number');
+          break;
+        } else {
+          lastData = possibleNumber;
+        }
       } else {
         throw new Error('unhandled signature');
       }
     }
-    shape.push(typeof data);
     return shape;
-  }
+  },
 
-  static addKeys(value, table) {
-    if (Array.isArray(value)) return;
+  addKeys(value: number[] | INumberHash, table: INumberHash): INumberHash {
+    if (Array.isArray(value)) return table;
     table = table || {};
```

Comment:
We might not need this line, as `table` is required and can't be null. so no need to re-assign.

Reply:
Ty sir.  And done.